blog-content/the-complete-guide-to-opentelemetry-in-next-js.md

```diff
@@ -0,0 +1,394 @@
+
+```typescript
+import { registerOTel } from "@vercel/otel";
+import { SimpleSpanProcessor } from "@opentelemetry/sdk-trace-node";
+import { PeriodicExportingMetricReader } from "@opentelemetry/sdk-metrics";
+import { SimpleLogRecordProcessor } from "@opentelemetry/sdk-logs";
+import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-grpc";
+import { OTLPMetricExporter } from "@opentelemetry/exporter-metrics-otlp-grpc";
+import { OTLPLogExporter } from "@opentelemetry/exporter-logs-otlp-grpc";
+
+export function register() {
+  if (process.env.NEXT_RUNTIME === "nodejs") {
+    registerOTel({
+      serviceName: "my-next-js-app",
+      attributes: {
+        "highlight.project_id": "<YOUR_PROJECT_ID>",
+        "deployment.environment.name": process.env.NODE_ENV,
+      },
+      propagators: ["auto"],
+      spanProcessors: [
+        new SimpleSpanProcessor(
+          new OTLPTraceExporter({
+            url: "https://otel.highlight.io:4317/v1/traces",
+            keepAlive: true,
+          }),
+        ),
+      ],
+      logRecordProcessor: new SimpleLogRecordProcessor(
+        new OTLPLogExporter({
+          url: "https://otel.highlight.io:4317/v1/logs",
+          keepAlive: true,
+        }),
+      ),
+      metricReader: new PeriodicExportingMetricReader({
+        exporter: new OTLPMetricExporter({
+          url: "https://otel.highlight.io:4317/v1/metrics",
+          keepAlive: true,
+        }),
+      }),
+    });
+  }
+}
+```
+
+Now, let's use the OpenTelemetry SDK in our route to emit data:
+
+```typescript
+import { NextResponse } from "next/server";
+import api, { propagation } from "@opentelemetry/api";
+import { logs, SeverityNumber } from "@opentelemetry/api-logs";
+
+export async function GET() {
+  const { email, name } = req.query;
+
+  // Here, you would typically make a call to your Python service
+  // passing the user's email or other identifier
+  // For now, we'll just log the information and return mock data
+  console.log(`Fetching data for user: ${email} ${name}`);
+
+  // In a real implementation, you would fetch data from the Garmin Health API here
+  const tracerProvider = api.trace.getTracerProvider();
+  const tracer = tracerProvider.getTracer("data");
+  const data = await tracer.startActiveSpan(
+    "data.fetch",
+    {
+      attributes: {
+        "user.email": email || undefined,
+        "user.name": name || undefined,
+      },
+    },
+    async () => {
+      console.log("Fetching data...", { email });
+      const headers = {
+        "Content-Type": "application/json",
+        Authorization: `Basic ${process.env.PYTHON_API_SECRET}`,
+      };
+      propagation.inject(api.context.active(), headers);
+      const response = await fetch(
+        `${process.env.PYTHON_API_HOSTNAME}/users/scrape`,
```

Comment:
Python...

Reply:
haha, this was my example app where i'm calling a python backend from my next.js backend (totally valid). will remove the env var tho and add some comments to make this more clear.